### ISSUE TITLE
cuda: MoE cache rejects zero-output Flash prompt batch with checkpoints enabled

### ISSUE THREAD

**Issue by GenerelSchwerz** [GenerelSchwerz/llama.cpp#78]
## Summary

A five-token Flash Next prompt fails before generation when context checkpoints are enabled. Changing only `--ctx-checkpoints 32` to `--ctx-checkpoints 0` makes the same request succeed.

Confirmed on fork commit `d7b6b6d0e3ad7c1de49b26911875d1c86bafab9f` (build 10904). The earlier failing log used `2a01b8834`; the intervening commit changes only Windows tests, not inference code.

## Environment and reproduction

- Windows build 26100.8457, RTX 5090, i9-14900K, 128 GiB RAM.
- CUDA 13.2, MSVC 19.50.35730.0, Release build, CUDA architecture `120a-real`.
- Model: `Qwen3.8-Flash-Next-UD-Q3_K_XL`, three-shard GGUF.

Set `LLAMA_ATTN_ROT_DISABLE=1` in the server environment and launch:

```text
llama-server.exe -m <model-directory>/Qwen3.8-Flash-Next-UD-Q3_K_XL-00001-of-00003.gguf --alias flash-next --host 127.0.0.1 --port 18797 -ngl all --jinja -c 8192 -np 1 -b 2048 -ub 256 -t 12 -tb 12 --spec-type none --load-mode none --lazy-mode on -fit off -fa on -ctk q8_0 -ctv q8_0 --moe-expert-cache-size 80 --cache-ram 0 --ctx-checkpoints 32 --no-warmup --experimental-logs -lv 5
```

After `/health` returns 200, POST this body to `/completion`:

```json
{"prompt":[7734,264,2716,11321,40719],"n_predict":8,"temperature":0,"top_k":20,"top_p":0.95,"min_p":0.05,"repeat_penalty":1,"seed":12345,"cache_prompt":false,"stream":false}
```

The prompt is the first five tokens from tokenizing `Write a short friendly greeting.` with `add_special=false`.

## Results

| Checkpoints | Result |
| --- | --- |
| `32` | HTTP 500, no generated tokens; exact error below |
| `0` | HTTP 200, 5 evaluated prompt tokens and all 8 requested generated tokens; no certificate errors |

```text
moe-cache: grouped decode certificate failed: semantic_group=47 layout=1 domain=1 tensor=blk.47.ffn_gate_exps.weight graph=geometry(5)
```

The two cases used identical request bodies, environment and server arguments except the checkpoint count. Both loaded in approximately 63 seconds. Available RAM stayed above 34 GiB and no resource guard fired. Both servers were stopped after collection. This is a correctness probe, not a throughput benchmark.

## Diagnosis

The [server checkpoint split](https://github.com/GenerelSchwerz/llama.cpp/blob/d7b6b6d0e3ad7c1de49b26911875d1c86bafab9f/tools/server/server-context.cpp#L3748-L3792) stops four tokens before prompt end. For a five-token prompt, the first batch therefore contains one token with no requested output logits. The failing log confirms that only one prompt token was queued.

Flash [prunes final-layer output rows](https://github.com/GenerelSchwerz/llama.cpp/blob/d7b6b6d0e3ad7c1de49b26911875d1c86bafab9f/src/models/qwen4exp.cpp#L402-L408), producing empty final-layer MoE geometry. The cached-reader inventory still includes these readers, while [geometry validation](https://github.com/GenerelSchwerz/llama.cpp/blob/d7b6b6d0e3ad7c1de49b26911875d1c86bafab9f/ggml/src/ggml-cuda/moe-cache.cu#L1657-L1703) requires positive ID dimensions/materialized IDs. This explains the layer-47 rejection. An initial `main/do_checkpoint=no` debug message does not rule out this split: the later check suppresses saving an initial checkpoint after the batch has already been split.

Expected: valid non-final prompt batches requesting no output logits should process successfully, while retaining certificate validation for actual expert consumers.

Workaround for this reproduced path: `--ctx-checkpoints 0`. No production fix or certificate bypass was applied. Full commands, requests, responses, logs and the reproduction script are preserved locally and on the test machine; private paths and access details are omitted here.

AI assistance: reproduction, source analysis and this report were prepared with Codex.

**Comment by GenerelSchwerz** [comment]
Resolved on the current `moe-cache` lineage.

The production behavior was fixed by `15e501cb0`, which excludes empty `MUL_MAT_ID` nodes from grouped MoE coverage inventory, plan inventory, and reader observation. That commit was not present in the originally reported `d7b6b6d0` lineage. Real non-empty expert consumers still go through certificate and geometry validation.

Regression coverage is now on `fix/issue-78-zero-output-checkpoints` at `4200ef78e`.

Verification on current head:
- Exact Qwen3.8 Flash Next Q3_K_XL five-token request with `--ctx-checkpoints 32`: HTTP 200, 5 prompt tokens evaluated, 8 tokens generated, no certificate failure.
- The log confirmed the one-token checkpoint split and omitted empty semantic group 47 from that graph. Later real-output graphs included group 47 normally.
- `--ctx-checkpoints 0` returned identical content and token counts.
- Full `test-moe-cache` suite passed.
- `test-moe-cache --grouped-layer-only` passed.

Hardware boundary: this revalidation used Linux, CUDA 13.4, and one RTX 5070 Ti with the expert cache reduced to 8 slots. Windows, RTX 5090, cache size 80, and multi-GPU execution were not revalidated.

AI assistance: investigation and regression coverage were prepared with Codex.
